sw/otbn/crypto/p256.s

```diff
@@ -1063,8 +1075,17 @@ scalar_mult_int:
   bn.addi   w9, w31, 1
   bn.mov    w10, w31
 
+  /* Shift shares of k so their MSBs are in the most significant position of a
+     word.
+       w0,w1 <= [w0, w1] << 192 = k0 << 192
+       w2,w3 <= [w2, w3] << 192 = k1 << 192 */
+  bn.rshi   w1, w1, w0 >> 64
```

Comment:
Just for my understanding, is `rshi` practically speaking a circular rotate operation? I am trying to understand how the comment relates to the instruction.

Reply:
Pretty much, yes. You concatenate the two source registers and then shift them right, and then you write the lower 256 bits of the result to the destination register. So:
```asm
bn.rshi   w1, w2, w3 >> 5
```
translates to:
```
w1 = ((w2 << 256 | w3) >> 5) mod 2^256
```
Practically, this means you can use it as a rotate, a right shift, or a left shift (using the `w31` all-zero register).
```
bn.rshi   w1, w31, w2 >> 5    // right shift: w2 >> 5
bn.rshi   w1, w2, w31 >> 251  // left shift: w2 << 5
bn.rshi   w1, w2, w2 >> 5     // rotate: rotr(w2, 5)
```